circuit/program/src/data/future/equal.rs

```diff
@@ -0,0 +1,45 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+impl<A: Aleo> Equal<Self> for Future<A> {
+    type Output = Boolean<A>;
+
+    /// Returns `true` if `self` and `other` are equal.
+    fn is_equal(&self, other: &Self) -> Self::Output {
+        // Recursively check each argument for equality.
+        let mut equal = Boolean::constant(true);
+        for (argument_a, argument_b) in self.arguments.iter().zip_eq(other.arguments.iter()) {
+            equal &= argument_a.is_equal(argument_b);
+        }
+
+        // Check the `program_id`, and `function_name`.
+        self.program_id.is_equal(&other.program_id) & equal & self.function_name.is_equal(&other.function_name)
```

Comment:
I'm assuming all of these operations need to be performed, hence the use of `&` and no short-circuiting?

Reply:
checked another PR related to circuits, the answer is "yes"